files/ebs_volume.sh

```diff
@@ -0,0 +1,130 @@
+#!/usr/bin/env bash
+
+set -euxo pipefail
+
+until ping -c 1 google.com &> /dev/null; do
+  echo "waiting for outbound connectivity"
+  sleep 5
+done
+
+# Set common variables used throughout the script.
+imds_token=$( curl -Ss -H "X-aws-ec2-metadata-token-ttl-seconds: 300" -XPUT 169.254.169.254/latest/api/token )
+instance_id=$( curl -Ss -H "X-aws-ec2-metadata-token: $imds_token" 169.254.169.254/latest/meta-data/instance-id )
+availability_zone=$( curl -Ss -H "X-aws-ec2-metadata-token: $imds_token" 169.254.169.254/latest/meta-data/placement/availability-zone )
+volume_id=""
+
+# Search for an available EBS volume to attach to the instance. Wait one minute for a volume to become available,
+# if no volume is found - create new one, attach, format and mount the volume.
+
+for i in $(seq 1 6); do
+
+  volume_id=$(
+    aws --cli-connect-timeout 300 ec2 describe-volumes \
+      --filters "Name=status,Values=available" "Name=availability-zone,Values=$availability_zone" "Name=tag:Name,Values=${name}-graphdb-data" \
+      --query "Volumes[*].{ID:VolumeId}" \
+      --output text | \
+      sed '/^$/d'
+  )
+  
+  if [ -z "${volume_id:-}" ]; then
+    echo 'ebs volume not yet available'
+    sleep 10
+  else
+    break
+  fi
+done
+
+if [ -z "${volume_id:-}" ]; then
+
+throughput_option=""
+if [ "${ebs_volume_type}" = "gp3" ] || [ "${ebs_volume_type}" = "gp2" ]; then
+    throughput_option="--throughput ${ebs_volume_throughput}"
+else
+    throughput_option=""
+fi
+
+volume_id=$(
+        aws --cli-connect-timeout 300 ec2 create-volume \
+          --availability-zone "$availability_zone" \
+          --encrypted \
+          --kms-key-id "${ebs_kms_key_arn}" \
+          --volume-type "${ebs_volume_type}" \
+          --size "${ebs_volume_size}" \
+          --iops "${ebs_volume_iops}" \
+          ${throughput_option:-} \
+          --tag-specifications "ResourceType=volume,Tags=[{Key=Name,Value=${name}-graphdb-data}]" | \
+          jq -r .VolumeId
+  )        
+
+  aws --cli-connect-timeout 300 ec2 wait volume-available --volume-ids "$volume_id"
+fi
+
+aws --cli-connect-timeout 300 ec2 attach-volume \
+  --volume-id "$volume_id" \
+  --instance-id "$instance_id" \
+  --device "${device_name}"
+
+# Handle the EBS volume used for the GraphDB data directory
+# beware, here be dragons...
+# read these articles:
+# https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/device_naming.html
+# https://github.com/oogali/ebs-automatic-nvme-mapping/blob/master/README.md
+
+# this variable comes from terraform, and it's what we specified in the launch template as the device mapping for the ebs
```

Comment:
@yaskoo any ideas how to handle the `device_mapping`, we don't have terraform here. Probably should be provided in an env file.